lib/replicache.dart

```diff
@@ -192,8 +230,102 @@ class Replicache implements ReadTransaction {
   }) =>
       query((tx) => tx.scan(prefix: prefix, start: start, limit: limit));
 
-  /// Synchronizes the database with the server. New local transactions that have been executed since the last
-  /// sync are sent to the server, and new remote transactions are received and replayed.
+  Future<void> _sync() async {
+    await _opened;
+    if (_closed) {
+      return;
+    }
+    final syncHead = await _beginSync();
+    await _maybeEndSync(syncHead);
+  }
+
+  Future<String> _beginSync() async {
+    await _opened;
+    final beginSyncResult = await _invoke(_name, 'beginSync', {
+      'batchPushURL': _batchUrl,
+      // TODO(arv): Rename this param.
+      'clientViewURL': _remote,
+      'dataLayerAuth': _clientViewAuth,
+    });
+
+    final syncInfo = beginSyncResult['syncInfo'];
+
+    void checkStatus(Map<String, dynamic> data, String serverName) {
+      final httpStatusCode = data['httpStatusCode'];
+      switch (httpStatusCode) {
+        case HttpStatus.ok:
+          break;
+        case HttpStatus.unauthorized:
+          // TODO: Reauth
+          break;
+        default:
+          print(
+              'Got unexpected response from $serverName server: $httpStatusCode: ${data['errorMessage']}');
+      }
+    }
+
+    final batchPushInfo = syncInfo['batchPushInfo'];
+    if (batchPushInfo != null) {
+      checkStatus(batchPushInfo, 'batch');
+      final mutationInfos = batchPushInfo['batchPushResponse']['mutationInfos'];
+      if (mutationInfos != null) {
+        for (final mutationInfo in mutationInfos) {
+          print(
+              'MutationInfo: ID: ${mutationInfo['id']}, Error: ${mutationInfo['error']}');
+        }
+      }
+    }
+
+    checkStatus(syncInfo['clientViewInfo'], 'client view');
+
+    return beginSyncResult['syncHead'];
+  }
+
+  Future<void> _maybeEndSync(String syncHead) async {
+    await _opened;
+    if (_closed) {
+      return;
+    }
+    final res = await _invoke(_name, 'maybeEndSync', {'syncHead': syncHead});
+    final replayMutations = res['replayMutations'];
+    if (replayMutations == null || replayMutations.isEmpty) {
+      // All done.
+      await _checkChange(syncHead);
```

Comment:
super newb to this code but this might be a bug. we expect root to be equal to synchead instantly after maybeendsync succeeds, so even though head changed this call to checkChange will not fire onchange.

Reply:
I think it's right.

The bindings don't monitor the database directly, the only way they know that the master changed is when either tx commits or sync lands, and tells it that master (may have) changed.

Reply:
See: https://github.com/rocicorp/replicache-sdk-flutter/blob/master/lib/replicache.dart#L277

Reply:
hmm, i still don't understand. on line 293 we are passing synchead to checkchange. since sync just completed we expect synchead to be equal to head/root so line 277 will be false and it will not fire onchange. what's the intent of doing a check like that?

Reply:
There is an issue here. We should not check for changes until the end of sync. Right now we check for the intermediate commits.